### PR TITLE
Use [Conformance] to tag conformance tests

### PR DESCRIPTION
**Pull request by sttts**
This PR removes the blacklist in hack/conformance-test.sh and replaces it with tagging in the description of the conformance tests themselves. I.e. conformance tests are selected using a white list.

TODO:
- [x] compare with previous master conformance test selection
- [x] compare with v1 test selection and create a diff for review
- [x] get somebody to review the new conformance tests compared to v1

Changes from conformance-v1 to this PR:

``` diff
--- filter-conformance-v1.txt   2015-10-07 20:46:17.000000000 +0100
+++ sorted-conformance-tag.txt  2015-10-08 14:39:19.000000000 +0100
@@ -3,18 +3,44 @@
 Docker Containers should be able to override the image's default commmand (docker entrypoint)
 Docker Containers should use the image defaults if command and args are blank
 Downward API should provide pod name and namespace as env vars
-EmptyDir volumes should have the correct mode
-EmptyDir volumes should support r/w
+Downward API volume should provide labels and annotations files
+EmptyDir volumes should support (non-root,0644,default)
+EmptyDir volumes should support (non-root,0644,tmpfs)
+EmptyDir volumes should support (non-root,0666,default)
+EmptyDir volumes should support (non-root,0666,tmpfs)
+EmptyDir volumes should support (non-root,0777,default)
+EmptyDir volumes should support (non-root,0777,tmpfs)
+EmptyDir volumes should support (root,0644,default)
+EmptyDir volumes should support (root,0644,tmpfs)
+EmptyDir volumes should support (root,0666,default)
+EmptyDir volumes should support (root,0666,tmpfs)
+EmptyDir volumes should support (root,0777,default)
+EmptyDir volumes should support (root,0777,tmpfs)
+EmptyDir volumes volume on default medium should have the correct mode
+EmptyDir volumes volume on tmpfs should have the correct mode
 Events should be sent by kubelets and the scheduler about pods scheduling and running
+Examples e2e [Example]ClusterDns should create pod that uses dns
 Kubectl client Guestbook application should create and stop a working application
+Kubectl client Kubectl api-versions should check if v1 is in available api versions
+Kubectl client Kubectl cluster-info should check if Kubernetes master services is included in cluster-info
+Kubectl client Kubectl describe should check if kubectl describe prints relevant information for rc and pods
+Kubectl client Kubectl expose should create services for rc
+Kubectl client Kubectl label should update the label on a resource
+Kubectl client Kubectl logs should be able to retrieve and filter logs
+Kubectl client Kubectl patch should add annotations for pods in rc
+Kubectl client Kubectl run pod should create a pod from an image when restart is Never
+Kubectl client Kubectl run pod should create a pod from an image when restart is OnFailure
+Kubectl client Kubectl run rc should create an rc from an image
+Kubectl client Kubectl version should check is all data is printed
+Kubectl client Proxy server should support --unix-socket=/path
+Kubectl client Proxy server should support proxy with --port 0
 Kubectl client Update Demo should create and stop a replication controller
 Kubectl client Update Demo should do a rolling update of a replication controller
 Kubectl client Update Demo should scale a replication controller
 Networking should function for intra-pod communication
 Networking should provide Internet connection for containers
 Networking should provide unchanging, static URL paths for kubernetes api services
+Pods should *not* be restarted with a /healthz http liveness probe
 Pods should *not* be restarted with a docker exec "cat /tmp/health" liveness probe
 Pods should be restarted with a /healthz http liveness probe
 Pods should be restarted with a docker exec "cat /tmp/health" liveness probe
@@ -23,6 +49,10 @@
 Pods should be updated
 Pods should contain environment variables for services
 Pods should get a host IP
+Pods should have monotonically increasing restart count
+Port forwarding With a server that expects a client request should support a client that connects, sends data, and disconnects
+Port forwarding With a server that expects a client request should support a client that connects, sends no data, and disconnects
+Port forwarding With a server that expects no client request should support a client that connects, sends no data, and disconnects
 PreStop should call prestop when killing a pod
 Probing container with readiness probe should not be ready before initial delay and never restart
 Probing container with readiness probe that fails should never be ready and never restart
@@ -30,17 +60,19 @@
 Proxy version v1 should proxy logs on node with explicit kubelet port
 Proxy version v1 should proxy through a service and a pod
 Proxy version v1 should proxy to cadvisor
 ReplicationController should serve a basic image on each replica with a public image
+SchedulerPredicates validates resource limits of pods that are allowed to run
+SchedulerPredicates validates that NodeSelector is respected if matching
+SchedulerPredicates validates that NodeSelector is respected if not matching
 Secrets should be consumable from pods
 Service endpoints latency should not be very high
 ServiceAccounts should mount an API token into pods
 Services should provide secure master service
 Services should serve a basic endpoint from pods
 Services should serve multiport endpoints from pods
 Variable Expansion should allow composing env vars into new env vars
 Variable Expansion should allow substituting values in a container's args
 Variable Expansion should allow substituting values in a container's command
-hostDir should give a volume the correct mode
-hostDir should support r/w on tmpfs
+hostPath should give a volume the correct mode
+hostPath should support r/w
+kube-ui should check that the kube-ui instance is alive
```

### CODE REVIEW

**Comment by mikedanese** [comment]
cc @erictune

**Comment by sttts** [comment]
Compared to the list in https://github.com/kubernetes/kubernetes/issues/14763#issuecomment-144637924 (which was a few days old):
```diff
--- sorted-conformance.txt	2015-10-07 16:24:20.000000000 +0200
+++ sorted-conformance-tag.txt	2015-10-07 16:55:59.000000000 +0200
@@ -1,6 +1,8 @@
-Daemon set should launch a daemon pod on every node of the cluster
+Daemon set should run and stop complex daemon
+Daemon set should run and stop simple daemon
 Deployment deployment should create new pods
 Deployment deployment should delete old pods and create new ones
+Deployment deployment should scale up and down in the right order
 Docker Containers should be able to override the image's default arguments (docker cmd)
 Docker Containers should be able to override the image's default command and arguments
 Docker Containers should be able to override the image's default commmand (docker entrypoint)
@@ -22,30 +24,22 @@
 EmptyDir volumes volume on default medium should have the correct mode
 EmptyDir volumes volume on tmpfs should have the correct mode
 Events should be sent by kubelets and the scheduler about pods scheduling and running
-Examples e2e [Example]ClusterDns should create pod that uses dns
-KubeProxy should test kube-proxy
-Kubectl client Guestbook application should create and stop a working application
-Kubectl client Kubectl api-versions should check if v1 is in available api versions
-Kubectl client Kubectl cluster-info should check if Kubernetes master services is included in cluster-info
-Kubectl client Kubectl describe should check if kubectl describe prints relevant information for rc and pods
-Kubectl client Kubectl expose should create services for rc
-Kubectl client Kubectl label should update the label on a resource
-Kubectl client Kubectl logs should be able to retrieve and filter logs
-Kubectl client Kubectl patch should add annotations for pods in rc
-Kubectl client Kubectl run pod should create a pod from an image when restart is Never
-Kubectl client Kubectl run pod should create a pod from an image when restart is OnFailure
-Kubectl client Kubectl run rc should create an rc from an image
-Kubectl client Kubectl version should check is all data is printed
-Kubectl client Proxy server should support --unix-socket=/path
-Kubectl client Proxy server should support proxy with --port 0
-Kubectl client Update Demo should create and stop a replication controller
-Kubectl client Update Demo should do a rolling update of a replication controller
-Kubectl client Update Demo should scale a replication controller
-Kubelet resource usage tracking over 30m0s with 0 pods per node
-Kubelet resource usage tracking over 30m0s with 50 pods per node
+Guestbook application should create and stop a working application
+Kubectl api-versions should check if v1 is in available api versions
+Kubectl cluster-info should check if Kubernetes master services is included in cluster-info
+Kubectl describe should check if kubectl describe prints relevant information for rc and pods
+Kubectl expose should create services for rc
+Kubectl label should update the label on a resource
+Kubectl logs should be able to retrieve and filter logs
+Kubectl patch should add annotations for pods in rc
+Kubectl run pod should create a pod from an image when restart is Never
+Kubectl run pod should create a pod from an image when restart is OnFailure
+Kubectl run rc should create an rc from an image
+Kubectl version should check is all data is printed
 Networking should function for intra-pod communication
 Networking should provide Internet connection for containers
 Networking should provide unchanging, static URL paths for kubernetes api services
+Pods Pshould get a host IP
 Pods should *not* be restarted with a /healthz http liveness probe
 Pods should *not* be restarted with a docker exec "cat /tmp/health" liveness probe
 Pods should be restarted with a /healthz http liveness probe
@@ -54,13 +48,16 @@
 Pods should be submitted and removed
 Pods should be updated
 Pods should contain environment variables for services
+Pods should have monotonically increasing restart count
 PreStop should call prestop when killing a pod
 Probing container with readiness probe should not be ready before initial delay and never restart
 Probing container with readiness probe that fails should never be ready and never restart
-Proxy version v1 should proxy logs on node
-Proxy version v1 should proxy logs on node with explicit kubelet port
-Proxy version v1 should proxy through a service and a pod
-Proxy version v1 should proxy to cadvisor
+Proxy server should support --unix-socket=/path
+Proxy server should support proxy with --port 0
+Proxy should proxy logs on node nodeProxyTest(f, version, "/logs/") })
+Proxy should proxy logs on node with explicit kubelet port nodeProxyTest(f, version, ":10250/logs/") })
+Proxy should proxy through a service and a pod
+Proxy should proxy to cadvisor nodeProxyTest(f, version, ":4194/containers/") })
 ReplicationController should serve a basic image on each replica with a public image
 SchedulerPredicates validates resource limits of pods that are allowed to run
 SchedulerPredicates validates that NodeSelector is respected if matching
@@ -71,9 +68,15 @@
 Services should provide secure master service
 Services should serve a basic endpoint from pods
 Services should serve multiport endpoints from pods
+Update Demo should create and stop a replication controller
+Update Demo should do a rolling update of a replication controller
+Update Demo should scale a replication controller
 Variable Expansion should allow composing env vars into new env vars
 Variable Expansion should allow substituting values in a container's args
 Variable Expansion should allow substituting values in a container's command
+With a server that expects a client request should support a client that connects, sends data, and disconnects
+With a server that expects a client request should support a client that connects, sends no data, and disconnects
+With a server that expects no client request should support a client that connects, sends no data, and disconnects
 hostPath should give a volume the correct mode
 hostPath should support r/w
 kube-ui should check that the kube-ui instance is alive
```

**Comment by karlkfi** [comment]
ginkgo has a `-dryrun` flag you can use with `-v` to print out the names of all the tests it would run. That will get you all the describe/context prefixes, and if you use the same `-seed` then it'll always be in the same order.

**Comment by sttts** [comment]
Good hint. Didn't see that.

**Comment by sttts** [comment]
Here is the different between **conformance-v1** and **master**:
```diff
--- filter-conformance-v1.txt	2015-10-07 21:46:17.000000000 +0200
+++ sorted-conformance-master.txt	2015-10-08 00:38:46.000000000 +0200
@@ -1,20 +1,68 @@
+Daemon set should run and stop complex daemon
+Daemon set should run and stop simple daemon
+Deployment deployment should create new pods
+Deployment deployment should delete old pods and create new ones
+Deployment deployment should scale up and down in the right order
 Docker Containers should be able to override the image's default arguments (docker cmd)
 Docker Containers should be able to override the image's default command and arguments
 Docker Containers should be able to override the image's default commmand (docker entrypoint)
 Docker Containers should use the image defaults if command and args are blank
 Downward API should provide pod name and namespace as env vars
-EmptyDir volumes should have the correct mode
-EmptyDir volumes should support r/w
+Downward API volume should provide labels and annotations files
+EmptyDir volumes should support (non-root,0644,default)
+EmptyDir volumes should support (non-root,0644,tmpfs)
+EmptyDir volumes should support (non-root,0666,default)
+EmptyDir volumes should support (non-root,0666,tmpfs)
+EmptyDir volumes should support (non-root,0777,default)
+EmptyDir volumes should support (non-root,0777,tmpfs)
+EmptyDir volumes should support (root,0644,default)
+EmptyDir volumes should support (root,0644,tmpfs)
+EmptyDir volumes should support (root,0666,default)
+EmptyDir volumes should support (root,0666,tmpfs)
+EmptyDir volumes should support (root,0777,default)
+EmptyDir volumes should support (root,0777,tmpfs)
+EmptyDir volumes volume on default medium should have the correct mode
+EmptyDir volumes volume on tmpfs should have the correct mode
 Events should be sent by kubelets and the scheduler about pods scheduling and running
+Examples e2e [Example]ClusterDns should create pod that uses dns
+Job should keep restarting failed pods
+Job should run a job to completion when tasks sometimes fail and are locally restarted
+Job should run a job to completion when tasks sometimes fail and are not locally restarted
+Job should run a job to completion when tasks succeed
+Job should scale a job down
+Job should scale a job up
+Job should stop a job
 Kubectl client Guestbook application should create and stop a working application
+Kubectl client Kubectl api-versions should check if v1 is in available api versions
+Kubectl client Kubectl cluster-info should check if Kubernetes master services is included in cluster-info
+Kubectl client Kubectl describe should check if kubectl describe prints relevant information for rc and pods
+Kubectl client Kubectl expose should create services for rc
+Kubectl client Kubectl label should update the label on a resource
+Kubectl client Kubectl logs should be able to retrieve and filter logs
+Kubectl client Kubectl patch should add annotations for pods in rc
+Kubectl client Kubectl run pod should create a pod from an image when restart is Never
+Kubectl client Kubectl run pod should create a pod from an image when restart is OnFailure
+Kubectl client Kubectl run rc should create an rc from an image
+Kubectl client Kubectl version should check is all data is printed
+Kubectl client Proxy server should support --unix-socket=/path
+Kubectl client Proxy server should support proxy with --port 0
 Kubectl client Update Demo should create and stop a replication controller
 Kubectl client Update Demo should do a rolling update of a replication controller
 Kubectl client Update Demo should scale a replication controller
+Kubelet experimental resource usage tracking over 30m0s with 50 pods per node
+Kubelet regular resource usage tracking over 30m0s with 0 pods per node
+Kubelet regular resource usage tracking over 30m0s with 35 pods per node
 Networking should function for intra-pod communication
 Networking should provide Internet connection for containers
 Networking should provide unchanging, static URL paths for kubernetes api services
+Pods should *not* be restarted with a /healthz http liveness probe
 Pods should *not* be restarted with a docker exec "cat /tmp/health" liveness probe
 Pods should be restarted with a /healthz http liveness probe
 Pods should be restarted with a docker exec "cat /tmp/health" liveness probe
@@ -23,7 +71,11 @@
 Pods should be updated
 Pods should contain environment variables for services
 Pods should get a host IP
+Port forwarding With a server that expects a client request should support a client that connects, sends data, and disconnects
+Port forwarding With a server that expects a client request should support a client that connects, sends no data, and disconnects
+Port forwarding With a server that expects no client request should support a client that connects, sends no data, and disconnects
 PreStop should call prestop when killing a pod
 Probing container with readiness probe should not be ready before initial delay and never restart
 Probing container with readiness probe that fails should never be ready and never restart
 Proxy version v1 should proxy logs on node
@@ -32,15 +84,22 @@
 Proxy version v1 should proxy to cadvisor
 ReplicationController should serve a basic image on each replica with a private image
 ReplicationController should serve a basic image on each replica with a public image
+SchedulerPredicates validates resource limits of pods that are allowed to run
+SchedulerPredicates validates that NodeSelector is respected if matching
+SchedulerPredicates validates that NodeSelector is respected if not matching
 Secrets should be consumable from pods
 Service endpoints latency should not be very high
 ServiceAccounts should mount an API token into pods
-Services should correctly serve identically named services in different namespaces on different external IP addresses
 Services should provide secure master service
 Services should serve a basic endpoint from pods
 Services should serve multiport endpoints from pods
 Variable Expansion should allow composing env vars into new env vars
 Variable Expansion should allow substituting values in a container's args
 Variable Expansion should allow substituting values in a container's command
-hostDir should give a volume the correct mode
-hostDir should support r/w on tmpfs
+hostPath should give a volume the correct mode
+hostPath should support r/w
+kube-ui should check that the kube-ui instance is alive
```

Note: some tests are not skipped by the conformance-test.sh skip patterns, but via code, e.g. for SSH based tests or those for special providers. I have removed them from the upper list because in the conformance test situation they won't be run.

**Comment by sttts** [comment]
And this is the difference of conformance tests between **master** and **this PR**:
```diff
--- filter-conformance-master.txt	2015-10-08 00:46:38.000000000 +0200
+++ sorted-conformance-tag.txt	2015-10-08 00:55:12.000000000 +0200
@@ -25,15 +25,6 @@
 EmptyDir volumes volume on tmpfs should have the correct mode
 Events should be sent by kubelets and the scheduler about pods scheduling and running
 Examples e2e [Example]ClusterDns should create pod that uses dns
-Job should keep restarting failed pods
-Job should run a job to completion when tasks sometimes fail and are locally restarted
-Job should run a job to completion when tasks sometimes fail and are not locally restarted
-Job should run a job to completion when tasks succeed
-Job should scale a job down
-Job should scale a job up
-Job should stop a job
 Kubectl client Guestbook application should create and stop a working application
 Kubectl client Kubectl api-versions should check if v1 is in available api versions
 Kubectl client Kubectl cluster-info should check if Kubernetes master services is included in cluster-info
@@ -51,17 +42,11 @@
 Kubectl client Update Demo should create and stop a replication controller
 Kubectl client Update Demo should do a rolling update of a replication controller
 Kubectl client Update Demo should scale a replication controller
-Kubelet experimental resource usage tracking over 30m0s with 50 pods per node
 Kubelet regular resource usage tracking over 30m0s with 0 pods per node
 Kubelet regular resource usage tracking over 30m0s with 35 pods per node
 Networking should function for intra-pod communication
@@ -71,18 +56,17 @@
 Pods should be updated
 Pods should contain environment variables for services
 Pods should get a host IP
+Pods should have monotonically increasing restart count
 Port forwarding With a server that expects a client request should support a client that connects, sends data, and disconnects
 Port forwarding With a server that expects a client request should support a client that connects, sends no data, and disconnects
```

**Comment by ixdy** [comment]
what's the diff between **conformance-v1** and **this PR**?

**Comment by karlkfi** [comment]
Don't we want those Job tests included?

**Comment by sttts** [comment]
Basically both of the posted diffs added. I can upload it tomorrow if necessary.

**Comment by sttts** [comment]
There was a discussion about experimental APIs. Are jobs still experimental? Same applies to the daemon sets of course.

**Comment by karlkfi** [comment]
My mistake. Yes, Jobs are experimental. Here are the other experimental apis: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/experimental/types.go

We can probably save that for another PR tho.

**Comment by sttts** [comment]
I have removed the experimental apis from the conformance suite. 

@mikedanese @erictune @kubernetes/goog-testing: Who can review the new additional tests in the PR description above?

**Comment by karlkfi** [comment]
Do we actually need "approval" of the new set if it's just a subset of what's already in master?

If someone wants to change the set later they can make another PR.

This PR just switches from blacklist to whitelist by tagging and removes experimental tests from conformance.

**Comment by sttts** [comment]
Rebased to latest master.

**Comment by karlkfi** [comment]
@k8s-bot test this

File: hack/conformance-test.sh [hunk 1]
```diff
@@ -74,38 +74,10 @@ git show-ref | grep $HEAD_SHA | grep refs/tags
...
-#  SSH: optional feature.
-#  Addon\supdate: requires SSH
-#  Volumes: contained only skipped tests.
-#  Clean\sup\spods\son\snode: performance
-#  MaxPods\slimit\snumber\sof\spods: not sure why this wasn't working on GCE but it wasn't.
-#  Kubectl\sclient\sSimple\spod: not sure why this wasn't working on GCE but it wasn't
-#  DNS: not sure why this wasn't working on GCE but it wasn't
-export CONFORMANCE_TEST_SKIP_REGEX="Cadvisor|MasterCerts|Density|Cluster\slevel\slogging|Etcd\sfailure|Load\sCapacity|Monitoring|Namespaces.*seconds|Pod\sdisks|Reboot|Restart|Nodes|Scale|Services.*load\sbalancer|Services.*NodePort|Services.*nodeport|Shell|SSH|Addon\supdate|Volumes|Clean\sup\spods\son\snode|Skipped|skipped|MaxPods\slimit\snumber\sof\spods|Kubectl\sclient\sSimple\spod|DNS|Resource\susage\sof\ssystem\scontainers"
+# It runs a whitelist of tests: all tests which are flagged with [Conformance]
+# somewhere in the description (i.e. either in the Describe part or the It part).
```
Comment (mikedanese) [inline hack/conformance-test.sh]: In this comment can you specifiy I command I could use to get the current list of conformance tests using dryrun?

**Comment by mikedanese** [comment]
Pretty much LGTM, we can refine the selection later if this doesn't work. Please address comment and squash.

File: hack/conformance-test.sh (hunk 1 above)
Comment (sttts) [inline hack/conformance-test.sh]: done

**Comment by sttts** [comment]
Squashed.

**Comment by mikedanese** [comment]
@sttts can you please submit a cherrypick?

**Comment by sttts** [comment]
@jdef will do

**Comment by erictune** [comment]
@sttts thanks for this I think it is simpler.

**Pull request merged**